ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3SubsystemDefaultPoolWriteHandler.java

```diff
@@ -78,44 +81,63 @@ protected void recordCapabilitiesAndRequirements(OperationContext context, Attri
-        if (poolAttribute.getName().equals(DEFAULT_SLSB_INSTANCE_POOL) || model.hasDefined(DEFAULT_SLSB_INSTANCE_POOL)) {
-
-            // NOTE: the new value may contain an expression, so we need to resolve it first
-            ModelNode resolvedNewValue = new ModelNode();
-            try {
-                resolvedNewValue = context.resolveExpressions(newValue);
-
-                // register the new requirement
-                String newSLSBRequirementName = RuntimeCapability.buildDynamicCapabilityName(STRICT_MAX_POOL_CONFIG_CAPABILITY_NAME, resolvedNewValue.asString());
-                context.registerAdditionalCapabilityRequirement(newSLSBRequirementName, DEFAULT_SLSB_POOL_CONFIG_CAPABILITY_NAME, DEFAULT_SLSB_INSTANCE_POOL);
-            } catch(OperationFailedException ofe) {
-                // if the new value cannot be resolved, deregister the old value only (in the finally clause)
-                EjbLogger.ROOT_LOGGER.defaultPoolExpressionCouldNotBeResolved(DEFAULT_SLSB_INSTANCE_POOL, model.get(DEFAULT_SLSB_INSTANCE_POOL).asString());
-
-            } finally {
-                // de-register the old requirement
-                String oldSLSBRequirementName = RuntimeCapability.buildDynamicCapabilityName(STRICT_MAX_POOL_CONFIG_CAPABILITY_NAME, oldValue.asString());
-                context.deregisterCapabilityRequirement(oldSLSBRequirementName, DEFAULT_SLSB_POOL_CONFIG_CAPABILITY_NAME, DEFAULT_SLSB_INSTANCE_POOL);
+        if (poolAttribute.getName().equals(DEFAULT_SLSB_INSTANCE_POOL)) {
+            if (oldValue.isDefined()) {
+
+                // NOTE: the new value may contain an expression, so we need to resolve it first
+                try {
+                    if (newValue.isDefined()) {
+                        ModelNode resolvedNewValue = context.resolveExpressions(newValue);
+
+                        // register the new requirement
+                        String newSLSBRequirementName = RuntimeCapability.buildDynamicCapabilityName(STRICT_MAX_POOL_CONFIG_CAPABILITY_NAME, resolvedNewValue.asString());
+                        context.registerAdditionalCapabilityRequirement(newSLSBRequirementName, DEFAULT_SLSB_POOL_CONFIG_CAPABILITY_NAME, DEFAULT_SLSB_INSTANCE_POOL);
+                    }
+                } catch (OperationFailedException ofe) {
+                    // if the new value cannot be resolved, deregister the old value only (in the finally clause)
+                    EjbLogger.ROOT_LOGGER.defaultPoolExpressionCouldNotBeResolved(DEFAULT_SLSB_INSTANCE_POOL, model.get(DEFAULT_SLSB_INSTANCE_POOL).asString());
+                } finally {
+                    // de-register the old requirement
+                    String oldSLSBRequirementName = RuntimeCapability.buildDynamicCapabilityName(STRICT_MAX_POOL_CONFIG_CAPABILITY_NAME, oldValue.asString());
+                    context.deregisterCapabilityRequirement(oldSLSBRequirementName, DEFAULT_SLSB_POOL_CONFIG_CAPABILITY_NAME, DEFAULT_SLSB_INSTANCE_POOL);
```

Comment:
@chengfang I think if you execute a :write-attribute with the same value that was there, you end up removing the capability requirement set a few lines above. Perhaps you need some guard to check that the new and old requirement names are not the same?

Essentially AbstractWriteAttributeHandler doesn't do any comparison itself of the old and new values to check

Reply:
I added a check to compare old and new requirement name, and only de-register it if they differ. I did it for both default-slsb-instance-pool and default-mdb-instance-pool.